osdep/io.c

```diff
@@ -255,31 +372,162 @@ int mp_printf(const char *format, ...)
+        flags |= FILE_ATTRIBUTE_NORMAL;
+        break;
+    case _O_CREAT | _O_TRUNC:
+        disposition = CREATE_ALWAYS;
+        break;
+    case _O_CREAT | _O_EXCL:
+    case _O_CREAT | _O_EXCL | _O_TRUNC:
+        disposition = CREATE_NEW;
+        flags |= FILE_ATTRIBUTE_NORMAL;
+        break;
+    }
+
+    // Opening a named pipe as a file can allow the pipe server to impersonate
+    // mpv's process, which could be a security issue. Set SQOS flags, so pipe
+    // servers can only identify the mpv process, not impersonate it.
+    if (disposition != CREATE_NEW)
+        flags |= SECURITY_SQOS_PRESENT | SECURITY_IDENTIFICATION;
+
+    // Keep the same semantics for some MSVCRT-specific flags
+    if (oflag & _O_TEMPORARY) {
+        flags |= FILE_FLAG_DELETE_ON_CLOSE;
+        access |= DELETE;
     }
+    if (oflag & _O_SHORT_LIVED)
+        flags |= FILE_ATTRIBUTE_TEMPORARY;
+    if (oflag & _O_SEQUENTIAL)
+        flags |= FILE_FLAG_SEQUENTIAL_SCAN;
+    else if (oflag & _O_RANDOM)
+        flags |= FILE_FLAG_RANDOM_ACCESS;
+
+    // Open the Windows file handle
     wchar_t *wpath = mp_from_utf8(NULL, filename);
-    int res = _wopen(wpath, oflag, mode);
+    HANDLE h = CreateFileW(wpath, access, share, NULL, disposition, flags, NULL);
     talloc_free(wpath);
-    return res;
+    if (h == INVALID_HANDLE_VALUE) {
+        set_errno_from_lasterror();
+        return -1;
+    }
+
+    // Map the Windows file handle to a CRT file descriptor. Note: MSVCRT only
+    // cares about the following oflags.
+    oflag &= _O_APPEND | _O_RDONLY | _O_RDWR | _O_WRONLY;
+    int fd = _open_osfhandle((intptr_t)h, oflag);
+    if (fd < 0) {
+        CloseHandle(h);
+        return -1;
+    }
+
+    return fd;
 }
 
 int mp_creat(const char *filename, int mode)
 {
-    return open(filename, O_CREAT|O_WRONLY|O_TRUNC, mode);
+    return mp_open(filename, _O_CREAT | _O_WRONLY | _O_TRUNC, mode);
 }
 
 FILE *mp_fopen(const char *filename, const char *mode)
 {
-    wchar_t *wpath = mp_from_utf8(NULL, filename);
-    wchar_t *wmode = mp_from_utf8(wpath, mode);
-    FILE *res = _wfopen(wpath, wmode);
-    talloc_free(wpath);
-    return res;
+    if (!*mode) {
+        errno = EINVAL;
+        return NULL;
+    }
+
+    // Parse the mode argument. To be more POSIXy, 'b' flags are ignored.
+    const char *pos = mode + 1;
+    bool plus = false;
+    if (*pos == 'b')
+        pos++;
+    if (*pos == '+') {
+        pos++;
+        plus = true;
+        if (*pos == 'b')
+            pos++;
+    }
+    if (*pos) {
+        errno = EINVAL;
+        return NULL;
+    }
+
+    int oflag;
+    switch (*mode) {
+    case 'r':
+        oflag = plus ? _O_RDWR : _O_RDONLY;
+        break;
+    case 'w':
+        oflag = (plus ? _O_RDWR : _O_WRONLY) | _O_CREAT | _O_TRUNC;
+        break;
+    case 'a':
+        oflag = (plus ? _O_RDWR : _O_WRONLY) | _O_CREAT | _O_APPEND;
+        break;
+    default:
+        errno = EINVAL;
+        return NULL;
+    }
```

Comment:
Wouldn't it be simpler to strictly parse the mode flags as a set?

Also support for `e` (O_CLOEXEC) and `x` (exclusive mode) might be interesting.

Reply:
Oh, probably. I guess 'b', '+' and extensions like 'e' and 'x' can appear in any order? If so, this should probably be a loop/switch. I'm not sure e/O_CLOEXEC will be useful, because I think Windows file handles are non-inheritable by default (but 'e' could definitely be parsed and ignored like 'b'.) x/O_EXCL should definitely be useful.

Reply:
From the manpage is _looks_ like the characters can be freely mixed. mpv probably doesn't do that anywhere (nor does it use e/x yet), but would probably be nice to conform to the standard and be extensible.

Reply:
Well I just looked at glibc, and they seem to allow the flags in any order. They also ignore all unrecognised flags. Apart from that, their logic is similar to mine.
https://github.com/bminor/glibc/blob/aa95a2414e4f/libio/fileops.c#L225-L279

We should probably do this too. POSIX 2008 says mode strings other than the ones in the spec cause undefined behaviour, so I guess we can do what we like here.